torch/csrc/distributed/rpc/tensorpipe_agent.h

```diff
@@ -131,6 +131,17 @@ class TensorPipeAgent : public RpcAgent {
       uint64_t requestSize,
       const std::string& destWorkerName);
 
+  // When a request+response completes, we need to mark the future message as
+  // complete. However, if its timeout has already expired, it already has an
+  // error set. There is no atomic "test-and-set" way to mark a future complete
+  // only if it isn't yet. It does exist for errors (setErrorIfNeeded) but, even
+  // then, it ends up printing a log message, which may worry the user. To solve
+  // both issues we use a separate atomic flag to know the status of the future.
+  struct AtomicFutureMessage {
```

Comment:
Should we instead add a "test-and-set" API into the Future type? I would assume this is also a problem for ProcessGroup agent?

Reply:
As there are many PRs depends on this one and as this is localized change, I think we can get this one in once the lint error is gone. Could you please create an issue for adding "test-and-set" API to make sure we don't forget about it?

Reply:
@mrshenli @lw I tried implementing `markCompletedIfNeeded` similar to `setErrorIfNeeded` here: https://github.com/pytorch/pytorch/pull/39080. Maybe this API change could simplify this PR?

Reply:
FWIW, I don't think this is a problem for the PG agent as we have a `futures_` map and we remove from that map if the future times out. In that case, we don't mark the future as completed a 2nd time because it's not found in the map .

Reply:
I just added a comment on https://github.com/pytorch/pytorch/issues/39082 - I think we actually may not necessarily need this change if we modify the data structure for storing timeout candidates. We can still keep the "test-and-set" API as a generic utility.

Reply:
Another problem with `setErrorIfNeeded` is that it always prints a log message which warns that it had to discard an error. Currently, the RPC thread will eventually time out all futures, which means the above warning would be printed for all messages, which is very noisy. If we could find a way to suppress those warnings then we could definitely go for this fix.

In #39082 other solutions are proposed and I'll comment there. I'd be inclined to land this as it is and then fix it properly later, so I don't have to resolve conflicts in all the diffs above this one in the stack ;)